gix-fsck/src/lib.rs

```diff
@@ -0,0 +1,120 @@
+//! A library for performing object database integrity and connectivity checks
+#![deny(rust_2018_idioms)]
+
+use gix_hash::ObjectId;
+use gix_hashtable::HashSet;
+use gix_object::{Kind, tree::EntryMode};
+use gix_odb::{Handle, FindExt, pack::Find};
+
+pub struct ConnectivityCheck<'a, F>
+where
+    F: FnMut(&ObjectId, Kind)
+{
+    /// ODB handle to use for the check
+    db: &'a Handle,
+    /// Closure to invoke when a missing object is encountered
+    missing_cb: F,
+    /// Set of Object IDs already (or about to be) scanned during the check
+    oid_set: HashSet,
+    /// Single buffer for decoding objects from the ODB
+    /// This is slightly faster than allocating throughout the connectivity check (and reduces the memory requirements)
+    buf: Vec<u8>,
+}
+
+impl<'a, F> ConnectivityCheck<'a, F>
+where
+    F: FnMut(&ObjectId, Kind)
+{
+    /// Instantiate a connectivity check
+    pub fn new(db: &'a Handle, missing_cb: F) -> ConnectivityCheck<'a, F> {
+        ConnectivityCheck {
+            db,
+            missing_cb,
+            oid_set: HashSet::default(),
+            buf: Vec::new(),
+        }
+    }
+
+    /// Run the connectivity check on the provided commit object ID
+    /// - This will walk the trees and blobs referenced by the commit and verify they exist in the ODB
+    /// - Any objects previously encountered by this [`ConnectivityCheck`] instance will be skipped
+    /// - Any referenced blobs that are not present in the ODB will result in a call to the  `missing_cb`
+    /// - Missing commits or trees will currently result in panic
+    ///     - TODO: add support for missing trees
+    ///     - TODO: consider how to handle a missing commit (invoke `missing_cb`, or possibly return a Result?)
+    pub fn check_commit(&mut self, oid: &ObjectId) {
```

Comment:
Now that this functionality is in a crate, it needs to be fully tested.

However, if it's too early for that, it's fine to move it all into `gitoxide-core` which implies it's pre-crate experimentation stage that will only be tested by hand, by running it on repositories.

Reply:
I've been working on adding some tests for this crate and plan on getting those into this PR before moving it out of draft. It's taking me a bit longer than I expected, mainly because I'm trying to get the test fixtures to produce deterministic repositories (at least with the assumption that the host machine has a somewhat recent `git` version). I see you've done some similar things around the existing tests.

I have no objections to ditching this new crate and moving it to `gitoxide-core`, but either way, I'd really like to have tests that make sure I'm not breaking anything as I progress.

Reply:
Yes, tests are definitely a good idea even though they aren't a requirement while it's experimental, particularly if it would move to `gitoxide-core`. However, I think once you have added tests, the code should stay in the crate which, as a container for sanity checks of the structure itself, does make sense.

Reply:
I've added a few tests in my latest update that exercise full, blobless, and treeless clones. There could definitely be more added, but wanted to get something in front of you sooner rather than later so you could see if I've done anything wrong with the test setup (particularly with the test fixture generation - let me know if I went off into the weeds there).